sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala

```diff
@@ -1079,6 +1079,17 @@ val GPU_COREDUMP_PIPE_PATTERN = conf("spark.rapids.gpu.coreDump.pipePattern")
     .booleanConf
     .createWithDefault(true)
 
+  val ENABLE_EXPAND_PREPROJECT = conf("spark.rapids.sql.expandPreproject.enabled")
```

Comment:
Introducing an extra configuration will be some extra usage burden. Could we have a check util method to figure out the existence of common expression. If we do have a common subexpression, we will intro a pre-project ahead of Expand node. Otherwise, we do nothing there.

It seems we can do this via a dry-run of tiered project. If duplicated ref exists in ```seq[seq[expression]]```, we will introduce a pre-project node to evaluate that duplicated refs.
https://github.com/NVIDIA/spark-rapids/blob/5d08aec8350dfe72f35ef0586a8aec97f1ef853c/sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala#L492-L507

Reply:
We can change the default value to `true` to eliminate user's extra attention. But I do not want to do this until we run enough tests to prove this will not lead to any regressions, especially for high GPU memory pressure cases.

>figure out the existence of common expression. If we do have a common subexpression, we will intro a pre-project ahead of Expand node. Otherwise, we do nothing there.

I guess this check `boundPreprojections.exprTiers.size > 1` will do as what you suggested.

Reply:
For me it all comes down to heuristics and our confidence in them.

If we don't have high confidence in the code, then we should have a config to be able to disable it as we try and run with it, but there should be a follow on issue to remove the config once we feel better about it. If we have really high confidence in the solution, then we don't need any kind of config to enable it or disable it. If we have such low confidence in it that we don't want it on yet, then it is probably not ready to be checked in. That is fine if it is a big feature and we are going to keep working on it, but for a small feature like this it is not.

The next part is a runtime the heuristic that deicides when and/how to apply this feature. This is separate from the config, because the config is there to help us after the feature has been released. From a computation standpoint I don't see how this would ever be worse than what we have today. But I can envision cases where the total memory usage might be much higher with this feature than without it. But I don't know how realistic they are in practice. Like if we have lots and lots of distinct operations on computed columns, with only a very small amount of overlap between them.

I am happy to see that literal values are not materialized until later, also from what I can tell the initial project throws away any columns that are not going to be materialized in the final output. 

There are a few more things that we might be able to do to reduce the memory, but I think it ends up with us guessing at the size of various outputs and then doing some form of a bin packing problem to reduce the amount of memory used, which could be done as a follow on issue if we run into any problems with this.